src/mono/sample/iOS-NativeAOT/README.md

```diff
@@ -11,29 +11,67 @@ This sample application is intended to be used by developers who work on enablin
 
 The sample shares the source code with the Mono sample specified at: `../iOS/Program.cs` and in general should have the same behavior as MonoAOT.
 
-## Limitations
+## Scenarios
 
-The application is **_currently_** relying on the following:
-1. Internal dependencies - locally building the internals is required as runtime and tools nuget packages are still not being produced
-2. Invariant globalization - `System.Globalization.Native` is currently not being built as part of NativeAOT framework for iOS-like platforms
-3. No publish targets - the SDK and MSBuild integration is still not complete
+There are two scenarios for building and testing the NativeAOT support for iOS:
+1. Testing against locally built internals
+    - Uses locally built ILCompiler, build integration targets, framework and runtime libraries
+    - Should be used in CI testing
+2. Testing against locally built packages - **end-to-end** testing
+    - References locally built ILCompiler and runtime packages
+    - Should be used by developers performing end-to-end validation
 
 ## How to build and test
 
-### Building for the first time
+### 1. Testing against locally built internals
 
 When building for the first time (on a clean checkout) run from this directory the following `make` command:
 ``` bash
 make world
 ```
-This will first build all required runtime components and dependencies, after which it will build the sample app and bundle it into an application bundle.
+
+The command performs the following:
+1. Build all required runtime components and dependencies
+2. Build the sample app and bundle it into an application bundle
+
 By default the build will use `Debug` build configuration and target `iossimulator`.
 To change this behavior, specify the desired setting in the following way:
 ``` bash
 make world BUILD_CONFIG=Release TARGET_OS=ios
 ```
 
-### To avoid building all the dependencies
+### 2. Testing against locally built packages - end-to-end testing
+
+When building for the first time (on a clean checkout) run from this directory the following `make` command:
+``` bash
+make world USE_RUNTIME_PACKS=true
+```
+
+The command performs the following:
+1. Builds ILCompiler and runtime packages:
+    - `Microsoft.DotNet.ILCompiler.8.0.0-dev` (host)
+    - `runtime.<host_os>-<host_arch>.Microsoft.DotNet.ILCompiler.8.0.0-dev` (host)
+    - `Microsoft.NETCore.App.Runtime.NativeAOT.<target_os>-<target_arch>.8.0.0-dev` (target)
+
+    NOTE:
+    - The packages can be found at: `artifacts/packages/<config>/Shipping/*.8.0.0-dev.nupkg`
+    - For testing incremental changes make sure to remove the **restored** nuget packages listed above with `8.0.0-dev` from your nuget restore directory (usually `~/.nuget/packages`). Failing to do so, can lead to unexpected behavior, as nuget will refuse to install newly generate package with a same version - `8.0.0-dev`. Something like:
```

Comment:
To avoid this problem, the other places in the repo recommend restoring the packages into a local directory using `--packages` option: https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/nativeaot.md#building-packages

Reply:
I agree and am aware of the problems it imposes. 
I did explore this option, but I had troubles with `--packages` in this setup, please take a look at my comment:  https://github.com/dotnet/runtime/pull/86652#issuecomment-1570227088 under `Smaller issue:`

Reply:
> The problem with this is that, during the build NuGetPackageRoot is already defined and points to the default ~/.nuget/packages/

`NuGetPackageRoot` is set very early to the directory specified using `--packages` option. What defined `NuGetPackageRoot` in this build?

Reply:
I am not sure how I can check what defined `NuGetPackageRoot` in the build. 
If I build the sample app with (from the sample directory): 
```bash
../../../../dotnet.sh \
        publish -c Release \
        -p:TargetOS=ios \
        -p:TargetArchitecture=arm64 \
        -p:DeployAndRun=true \
        -p:StripDebugSymbols=false \
        -p:PublishAotUsingRuntimePack=true \
        --packages packages \
        -bl
```
and inspect the `msbuild.binlog` in the Evaluation section these properties are being set:
```
NuGetPackageFolders = /Users/ivan/repos/runtime-mono-iOS/src/mono/sample/iOS-NativeAOT/packages
NuGetPackageRoot = /Users/ivan/.nuget/packages/
``` 
causing
```
Program.csproj id:25
        Property NuGetPackageRoot = /Users/ivan/.nuget/packages/
        NoImport $(NuGetPackageRoot)/microsoft.dotnet.ilcompiler/8.0.0-dev/build/Microsoft.DotNet.ILCompiler.props
```

---
Nevertheless, if before build I set the environment variable to: 
```
export NUGET_PACKAGES=/Users/ivan/repos/runtime-mono-iOS/src/mono/sample/iOS-NativeAOT/packages
```
then the build succeeds.

Should I make this workaround with env variable available in this PR change and potentially open a NuGet issue around `--packages` restore option?

Reply:
> I am not sure how I can check what defined NuGetPackageRoot in the build.

NuGetPackageFolders and NuGetPackageRoot properties are set here: https://github.com/NuGet/NuGet.Client/blob/20618ba68b3a6cd9928d52b4c5f7aa6872747ba4/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs#L174-L175

This code writes out `*.nuget.g.props` file during restore. This .props file is later included by the build - it is why you do not see anything interesting about NuGetPackageRoot in the binlog. The property was set in the included .props file and that's all.

To get to the bottom of this, check what is in the `*.nuget.g.props` file (I assume it is going to have the bad path too) and how it got there.

Reply:
> To get to the bottom of this, check what is in the *.nuget.g.props file (I assume it is going to have the bad path too) and how it got there.

That is what I am struggling with:
- as in `Program.csproj.nuget.g.props` the following is generated:
    ``` xml
    <NuGetPackageRoot Condition=" '$(NuGetPackageRoot)' == '' ">/Users/ivan/repos/runtime-mono-iOS/src/mono/sample/iOS-NativeAOT/packages</NuGetPackageRoot>
    <NuGetPackageFolders Condition=" '$(NuGetPackageFolders)' == '' ">/Users/ivan/repos/runtime-mono-iOS/src/mono/sample/iOS-NativeAOT/packages</NuGetPackageFolders>
    ...
    <Import Project="$(NuGetPackageRoot)/microsoft.dotnet.ilcompiler/8.0.0-dev/build/Microsoft.DotNet.ILCompiler.props"
    ```

- while in the msbuild.binlog it is stated that:
    ```
    NuGetPackageFolders = /Users/ivan/repos/runtime-mono-iOS/src/mono/sample/iOS-NativeAOT/packages
    NuGetPackageRoot = /Users/ivan/.nuget/packages/
    ... and further
    Import: /Users/ivan/repos/runtime-mono-iOS/artifacts/obj/mono/Program/Program.csproj.nuget.g.props at (62;3)
    ...
        NoImport: $(NuGetPackageRoot)/microsoft.dotnet.ilcompiler/8.0.0-dev/build/Microsoft.DotNet.ILCompiler.props at (22;5) false condition; (Exists('$(NuGetPackageRoot)/microsoft.dotnet.ilcompiler/8.0.0-dev/build/Microsoft.DotNet.ILCompiler.props') was evaluated as Exists('/Users/ivan/.nuget/packages//microsoft.dotnet.ilcompiler/8.0.0-dev/build/Microsoft.DotNet.ILCompiler.props')).
    ```
    
Which means that `NuGetPackageRoot` assignments in `Program.csproj.nuget.g.props` don't have any effect as the property is not empty at that point (already defined).

Reply:
I wonder if this is messing with how you want things set up. This file is sourced from dotnet.sh that you're running.

https://github.com/dotnet/runtime/blob/7febca6802545d8c96e386187d536bc100d40337/eng/common/tools.sh#L351

Reply:
Oh... that must be it! Thanks a bunch! 
I will adapt the makefile and README.md to use a local restore folder instead.

Reply:
This has now been fixed by using the environment variable: https://github.com/dotnet/runtime/pull/86652/files#diff-00aa52c8350c388c1879fdd89fce2ccb79848dedf2ee30827769b7adfdc64d3bR53-R55 
README.md has been updated accordingly.